src/backend/distributed/transaction/backend_data.c

```diff
@@ -1278,3 +1290,202 @@ DecrementExternalClientBackendCounter(void)
+	{
+		nextClusterClockValue = epochValue;
+	}
+
+	backendManagementShmemData->clusterClockValue = nextClusterClockValue;
+	SpinLockRelease(&backendManagementShmemData->clockMutex);
+	return nextClusterClockValue;
+}
+
+
+/*
+ * CurrentClusterClockValue returns the current logical clock value.
+ */
+uint64
+CurrentClusterClockValue(void)
+{
+	SpinLockAcquire(&backendManagementShmemData->clockMutex);
+	uint64 epochValue = backendManagementShmemData->clusterClockValue;
+	SpinLockRelease(&backendManagementShmemData->clockMutex);
+
+	return epochValue;
+}
+
+
+/*
+ * AdjustAndInitializeClusterClock compares the current epoch value with
+ * the persisted clock value to see if the clock drifted backwards, and
+ * adjusts the logical clock accordingly.
+ * Ideally, this routine should be called only _once_ at the time of boot.
+ */
+void
+AdjustAndInitializeClusterClock(void)
+{
+	uint64 epochPersistedValue;
+
+	if (!GetLocalClockValue(&epochPersistedValue))
+	{
+		/* couldn't get the value from catalog, might happened
+		 * during pg upgrade */
+		return;
+	}
+
+	/* Ensure clock never drifts back */
+	SpinLockAcquire(&backendManagementShmemData->clockMutex);
+
+	/* Check if the clock was already set after the boot */
+	if (backendManagementShmemData->clusterClockValue > 0)
+	{
+		/* Already initialized */
+		SpinLockRelease(&backendManagementShmemData->clockMutex);
+		return;
+	}
+
+	Assert(backendManagementShmemData->clusterClockValue == 0);
+	backendManagementShmemData->clusterClockValue = epochPersistedValue;
+
+	SpinLockRelease(&backendManagementShmemData->clockMutex);
+
+	uint64 epochValue = GetEpochTimeMs();
+	if (epochPersistedValue > epochValue)
+	{
+		ereport(LOG, (errmsg("Drift in the epoch clock, adjusted "
+							 "to persisted old value:(%ld)", epochPersistedValue)));
+	}
+}
+
+
+/*
+ * AdjustLocalClockToGlobal sets the node's local logical cluster clock value
+ * to the distributed transaction clock value, which is the maximum clock value
+ * of all the particpating nodes.
+ */
+void
+AdjustLocalClockToGlobal(void)
+{
+	if (!EnableGlobalClock || !MyBackendData)
+	{
+		return;
+	}
+
+	uint64 transactionClockValue = MyBackendData->transactionId.transactionClockValue;
+
+	/* It's ok to check unprotected as the real assignment is done under protection */
+	if (transactionClockValue <= backendManagementShmemData->clusterClockValue)
+	{
+		/*
+		 * Clock moved and greater than the rest of the nodes, there is
+		 * no need to adjust the clock.
+		 */
+		return;
+	}
+
+	SpinLockAcquire(&backendManagementShmemData->clockMutex);
+	backendManagementShmemData->clusterClockValue =
+		MyBackendData->transactionId.transactionClockValue;
+	SpinLockRelease(&backendManagementShmemData->clockMutex);
+}
+
+
+/*
+ * CurrentTransactionClockValue returns the clock value of the current
+ * distributed transaction. The caller must make sure a distributed
+ * transaction is in progress.
+ */
+uint64
+CurrentTransactionClockValue(void)
```

Comment:
what about GetTransactionIdClockValue to have similar naming to SetTransactionIdClockValue ?